### PR TITLE
Fix SwiftPackageManager copy rule parsing

### PR DESCRIPTION
**Pull request by alexanderwe**
Resolves https://github.com/tuist/tuist/issues/4724

### Short description 📝

This PR tries to resolves the issue of tuist not respecting the `.copy` rule of SPM resources. This is done by handling `.process` rules as it was done previously and for `.copy` rules a simple `.folderReference` is used. 

### How to test the changes locally 🧐

- Try to run `tuist fetch`  and `tuist generate` against https://github.com/alexanderwe/tuist-spm-resources-copy-issue and have a look at the generated project 
- The generated project should build fine without duplicated file errors

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [ ] The title of the PR will be used as changelog entry, please make sure it is clear and suitable
- [ ] In case the PR introduces changes that affect users, the documentation has been updated
- [ ] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`

### CODE REVIEW

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 1]
```diff
@@ -654,15 +654,19 @@ extension ResourceFileElements {
...
-            resources: resourcesPaths.map { absolutePath in
+        func handleProcessRule() -> [ProjectDescription.ResourceFileElement] {
+            let processResources = resources.filter { $0.rule == .process }
+
+            var resourcesPaths = processResources.map { path.appending(RelativePath($0.path)) }
+
+            if sources == nil {
+                // SPM automatically includes resources only if custom sources are not specified
+                resourcesPaths += defaultResourcePaths(from: path)
+            }
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:666]: This has nothing to do with `handleProcessRule`, I would handle it separately

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 2]
```diff
@@ -673,7 +677,24 @@ extension ResourceFileElements {
...
+            let copyResources = resources.filter { $0.rule == .copy }
+
+            let resourcesPaths = copyResources.map { path.appending(RelativePath($0.path)) }
+
+            return resourcesPaths.map { absolutePath in
+                return .folderReference(path: Path(absolutePath.pathString))
+            }
+        }
+
+        let resourceFileElements = handleProcessRule() +  handleCopyRule()
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:692]: I would rather have a single loop and switch over `$0.rule`, so we are sure we handle all scenarios

**Review by danyf90 — CHANGES_REQUESTED** [review]
Can you please add a test for the scenarios?

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift (hunk 2 above)
Comment (alexanderwe) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: Included a switch statement to ensure all scenarios are handled

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift (hunk 1 above)
Comment (alexanderwe) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: Moved handling of default resources out of `handleProcessRule` and handle them separately

**Comment by alexanderwe** [comment]
> Can you please add a test for the scenarios?

@danyf90 There were already some tests which included the `.copy` rule inside `PackageInfoMapperTests.swift`. I fixed them in order to reflect the changes that `.copy` rules is mapped to a `.folderReference`

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 3]
```diff
@@ -655,26 +655,55 @@ extension ResourceFileElements {
...
+        }
+
+        var resourceFileElements: [ProjectDescription.ResourceFileElement] = resources.map {
+            let resourceAbsolutePath = path.appending(RelativePath($0.path))
+
+            switch $0.rule {
+            case .copy:
+                return handleCopyResource(resourceAbsolutePath: resourceAbsolutePath)
+            case .process:
+                return handleProcessResource(resourceAbsolutePath: resourceAbsolutePath)
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:693]: nit: I would just put the logic here, without defining the additional functions

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 4]
```diff
@@ -655,26 +655,55 @@ extension ResourceFileElements {
...
-        )
+        }
+
+        // Add default resources path if necessary
+        // They are handled like a `.process` rule
+        if sources == nil {
+            resourceFileElements += defaultResourcePaths(from: path).map { handleProcessResource(resourceAbsolutePath: $0) }
+        }
+
+        // Sanity check
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:703]: I would not call it a sanity check, as it being empty is perfectly normal

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 5]
```diff
@@ -655,26 +655,55 @@ extension ResourceFileElements {
...
+        guard !resourceFileElements.isEmpty else { return nil }
+
+        return .init( resources: resourceFileElements)
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:706]:
```suggestion
        return .init(resources: resourceFileElements)
```

**Review by danyf90 — APPROVED** [review]
Could you please fix the tests? 🙏

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift (hunk 5 above)
Comment (alexanderwe) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: Fixed

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift (hunk 4 above)
Comment (alexanderwe) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: Good point, changed wording

**Review by danyf90 — APPROVED** [review]

**Comment by alexanderwe** [comment]
Just wanted to be sure - anything else I need to do here or just another approval missing ?

**Comment by danyf90** [comment]
@alexanderwe looks good to me, just waiting for another approval 🙏

**Comment by alexanderwe** [comment]
Alright 🙂

**Comment by Arafo** [comment]
I've just found that this change messes up projects using SPM dependencies where `.copy` rule points to a specific file instead of a folder path like here for example:
https://github.com/magicien/GLTFSceneKit/blob/a9587ec5c6515ee7d036533fdd909815a9a718ef/Package.swift#L27-L36

Leading to these warnings when running `tuist fetch`
```
Swift Package Manager dependencies installed successfully.
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/GLTFShaderModifierFragment_alphaCutoff.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/GLTFShaderModifierSurface.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/GLTFShaderModifierSurface_alphaModeBlend.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/GLTFShaderModifierSurface_doubleSidedWorkaround.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/KHR_materials_pbrSpecularGlossiness/GLTFShaderModifierSurface_pbrSpecularGlossiness.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/KHR_materials_pbrSpecularGlossiness/GLTFShaderModifierSurface_pbrSpecularGlossiness_doubleSidedWorkaround.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/KHR_materials_pbrSpecularGlossiness/GLTFShaderModifierSurface_pbrSpecularGlossiness_texture_doubleSidedWorkaround.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/VRM/GLTFShaderModifierFragment_VRMUnlitTexture.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/VRM/GLTFShaderModifierFragment_VRMUnlitTexture_Cutoff.shader is not a directory - folder reference paths need to point to directories
Tuist/Dependencies/SwiftPackageManager/.build/checkouts/GLTFSceneKit/Sources/GLTFSceneKit/Resources/VRM/GLTFShaderModifierSurface_VRMMToon.shader is not a directory - folder reference paths need to point to directories
```

And thus generated target does not contain these resources:
tuist_issue

Release 3.11.0 works fine:
tuist_3 11 0

Upload an example project also: https://github.com/Arafo/tuist-spm-resources-copy-issue

**Comment by alexanderwe** [comment]
@Arafo Thanks a lot for reporting this issue ! I am sorry that my PR causes a regression here. I think it is related to https://github.com/tuist/tuist/issues/4796. The copy rule parsing seems to fail for: 

* Single files that are copied 
* Opaque directories like .xcdatamodel which should be treated like files and not directories 

@danyf90 should we create a new issue here ?

**Pull request merged**